File: ruby/backend.rb
#!/usr/bin/env ruby
require "json"
require "aws-sdk-s3"
require "aws-sdk-dynamodb"

config_dir = File.join(__dir__, "../config")

puts "Setting up Terraform backends for all environments in #{config_dir}\n"

# Load shared.json once (terraform + aws data lives here now)
shared_file = File.join(config_dir, "shared.json")
shared_cfg = JSON.parse(File.read(shared_file))

bucket = shared_cfg["terraform"]["backend"]["bucket"]
table  = shared_cfg["terraform"]["backend"]["dynamodbTable"]
region = shared_cfg["aws"]["region"]

Dir.glob("#{config_dir}/*.json").each do |file|
  next if File.basename(file) == "shared.json" # skip shared.json itself

  env = File.basename(file, ".json")
  cfg = JSON.parse(File.read(file))

  puts "------------------------------------------------------------"
  puts "Environment:    #{env}"
  puts "Config file:    #{file}"
  puts "S3 Bucket:      #{bucket}"
  puts "DynamoDB Table: #{table}"
  puts "Region:         #{region}\n"

  creds = Aws::Credentials.new(
    ENV["AWS_ACCESS_KEY_ID"],
    ENV["AWS_SECRET_ACCESS_KEY"]
  )

  s3  = Aws::S3::Client.new(region: region, credentials: creds)
  ddb = Aws::DynamoDB::Client.new(region: region, credentials: creds)

  # --- S3 Bucket ---
  begin
    s3.head_bucket(bucket: bucket)
    puts "[OK] Bucket already exists: #{bucket}"
  rescue Aws::S3::Errors::NotFound, Aws::S3::Errors::NoSuchBucket
    puts "[CREATE] Bucket: #{bucket}"
    if region == "us-east-1"
      s3.create_bucket(bucket: bucket)
    else
      s3.create_bucket(bucket: bucket,
                       create_bucket_configuration: { location_constraint: region })
    end
  end

  # Enable versioning
  s3.put_bucket_versioning(bucket: bucket,
                           versioning_configuration: { status: "Enabled" })
  puts "[OK] Versioning enabled on bucket: #{bucket}"

  # --- DynamoDB Table ---
  begin
    ddb.describe_table(table_name: table)
    puts "[OK] DynamoDB table already exists: #{table}"
  rescue Aws::DynamoDB::Errors::ResourceNotFoundException
    puts "[CREATE] DynamoDB table: #{table}"
    ddb.create_table(
      table_name: table,
      attribute_definitions: [{ attribute_name: "LockID", attribute_type: "S" }],
      key_schema: [{ attribute_name: "LockID", key_type: "HASH" }],
      billing_mode: "PAY_PER_REQUEST"
    )
  end

  puts "[DONE] Backend ready for environment: #{env}\n"
end

puts "------------------------------------------------------------"
puts "All Terraform backends are set up."

File: go/backend.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	ddbTypes "github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3Types "github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/fatih/color"
)

type SharedConfig struct {
	Aws struct {
		Region string `json:"region"`
	} `json:"aws"`
	Terraform struct {
		Backend struct {
			Bucket        string `json:"bucket"`
			DynamoDBTable string `json:"dynamodbTable"`
		} `json:"backend"`
	} `json:"terraform"`
}

func main() {
	ctx := context.Background()
	title := color.New(color.FgCyan, color.Bold).SprintFunc()
	success := color.New(color.FgGreen).SprintFunc()
	warn := color.New(color.FgYellow).SprintFunc()
	fail := color.New(color.FgRed).SprintFunc()

	// Load shared.json
	sharedFile := "./config/shared/shared.json"
	data, err := os.ReadFile(sharedFile)
	if err != nil {
		log.Fatalf("Failed to read shared config: %v", err)
	}
	var cfg SharedConfig
	if err := json.Unmarshal(data, &cfg); err != nil {
		log.Fatalf("Invalid JSON in shared config: %v", err)
	}

	region := cfg.Aws.Region
	bucket := cfg.Terraform.Backend.Bucket
	table := cfg.Terraform.Backend.DynamoDBTable

	awsCfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(region))
	if err != nil {
		log.Fatalf("Unable to load AWS SDK config: %v", err)
	}

	s3Client := s3.NewFromConfig(awsCfg)
	dynamo := dynamodb.NewFromConfig(awsCfg)

	fmt.Println(title("==> Checking S3 bucket: " + bucket))
	_, err = s3Client.HeadBucket(ctx, &s3.HeadBucketInput{Bucket: aws.String(bucket)})
	if err == nil {
		fmt.Println(success("✔ Bucket already exists: " + bucket))
	} else {
		fmt.Println(warn("Creating bucket: " + bucket + "..."))
		if region == "us-east-1" {
			_, err = s3Client.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
		} else {
			_, err = s3Client.CreateBucket(ctx, &s3.CreateBucketInput{
				Bucket: aws.String(bucket),
				CreateBucketConfiguration: &s3Types.CreateBucketConfiguration{
					LocationConstraint: s3Types.BucketLocationConstraint(region),
				},
			})
		}
		if err != nil {
			log.Fatalf("%s Failed to create bucket: %v", fail("✖"), err)
		}
		fmt.Println(success("✔ Bucket created: " + bucket))
	}

	_, err = s3Client.PutBucketVersioning(ctx, &s3.PutBucketVersioningInput{
		Bucket: aws.String(bucket),
		VersioningConfiguration: &s3Types.VersioningConfiguration{
			Status: s3Types.BucketVersioningStatusEnabled,
		},
	})
	if err != nil {
		log.Fatalf("%s Failed to enable versioning: %v", fail("✖"), err)
	}
	fmt.Println(success("✔ Versioning enabled on bucket: " + bucket))

	fmt.Println(title("\n==> Checking DynamoDB table: " + table))
	_, err = dynamo.DescribeTable(ctx, &dynamodb.DescribeTableInput{TableName: aws.String(table)})
	if err == nil {
		fmt.Println(success("✔ Table already exists: " + table))
	} else {
		fmt.Println(warn("Creating DynamoDB table: " + table + "..."))
		_, err = dynamo.CreateTable(ctx, &dynamodb.CreateTableInput{
			TableName: aws.String(table),
			AttributeDefinitions: []ddbTypes.AttributeDefinition{
				{AttributeName: aws.String("LockID"), AttributeType: ddbTypes.ScalarAttributeTypeS},
			},
			KeySchema: []ddbTypes.KeySchemaElement{
				{AttributeName: aws.String("LockID"), KeyType: ddbTypes.KeyTypeHash},
			},
			BillingMode: ddbTypes.BillingModePayPerRequest,
		})
		if err != nil {
			log.Fatalf("%s Failed to create table: %v", fail("✖"), err)
		}
		fmt.Println(success("✔ Table created: " + table))
	}

	fmt.Println(success("\nTerraform backend setup complete!"))
	fmt.Println(color.CyanString("S3 Bucket: " + bucket))
	fmt.Println(color.CyanString("DynamoDB Table: " + table))
	fmt.Println(color.CyanString("Region: " + region))
}
